### DISCUSSION TITLE
How to include a directory provided on the command line?

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#4708] [Q&A]
I have tried passing the value of an `option` to `includes` but I can't make it work.  Is there a way of accessing option values at description scope?  We need user code to provide the path to where they have put our SDK, which contains the xmake.lua that describes how to build a project.  Currently, this ends up being hard coded in the project's xmake.lua and needs to be edited by hand, but I'm sure there's a better way of doing it.

**Comment by waruqi** [comment]
We can not use option value in includes directly, but you can try.

```lua
option("test", {description = ".."})

if has_config("test") then
    includes("test")
end
```

or

```lua
option("test", {description = ".."})

if is_config("test", "value") then
    includes("test")
end
```

or 


```lua
option("test", {description = ".."})

local value = get_config("test")
if value then
    includes("test" .. value)
end
```

**Reply by davidchisnall** [reply]
This looks like most of a solution.  The following works:

```lua
option("sdk-dir") do
	set_showmenu(true)
	set_description("The SDK directory")
	after_check(function (option)
		print("The SDK directory: ", option:value())
		if not option:value() then
			raise("The SDK directory must be provided with --sdk-dir")
		end
    end)
end

includes(path.absolute(get_config("sdk-dir")))
```

This would be simplified if `option` had a `set_required` that causes an error if it is not provided.  This would also be helpful for us with `--sdk`, since if you don't specify the toolchain location xmake defaults to one that doesn't work and then compiles fail.

**Reply by davidchisnall** [reply]
Unfortunately, that works in a small test case, but doesn't work in a real example.  It appears that the option is not evaluated sufficiently early, and so we don't get the includes (there's a warning that it's including `nil` appended to the include path) and we then get other errors about things after this depending on things that should be expoed from the `xmake.lua` from the SDK.  These are hard to read, because xmake prints its `--help` output after the error message, so you have to scroll up a lot to see the real cause.  

This is part of the undocumented multi-pass execution of xmake, which makes it very hard to reason about what will actually happen.

**Comment by waruqi** [comment]
Of course, the recommended approach is to include all scripts in the description scope, and then enable or disable the target or something else in the script scope.

**Reply by davidchisnall** [reply]
I'm not sure how that applies to the problem.
